pkg/scheduler/util/backoff_utils.go

```diff
@@ -59,35 +50,33 @@ func (b *backoffEntry) unlock() {
 	}
 }
 
-// TryWait tries to acquire the backoff lock, maxDuration is the maximum allowed period to wait for.
-func (b *backoffEntry) TryWait(maxDuration time.Duration) bool {
-	if !b.tryLock() {
-		return false
-	}
-	defer b.unlock()
-	b.wait(maxDuration)
-	return true
+func (b *backoffEntry) backoffTime() time.Time {
+	return b.lastUpdate.Add(b.backoff)
 }
 
-func (entry *backoffEntry) getBackoff(maxDuration time.Duration) time.Duration {
-	duration := entry.backoff
-	newDuration := time.Duration(duration) * 2
+func (entry *backoffEntry) doBackoff(maxDuration time.Duration) time.Duration {
+	if !entry.initialized {
+		entry.initialized = true
+		return entry.backoff
+	}
+	newDuration := entry.backoff * 2
 	if newDuration > maxDuration {
 		newDuration = maxDuration
 	}
 	entry.backoff = newDuration
-	glog.V(4).Infof("Backing off %s", duration.String())
-	return duration
+	glog.V(4).Infof("Backing off %s", newDuration.String())
+	return newDuration
 }
 
-func (entry *backoffEntry) wait(maxDuration time.Duration) {
-	time.Sleep(entry.getBackoff(maxDuration))
+func (entry *backoffEntry) backoffAndWait(maxDuration time.Duration) {
+	time.Sleep(entry.doBackoff(maxDuration))
 }
 
 type PodBackoff struct {
-	perPodBackoff   map[ktypes.NamespacedName]*backoffEntry
+	expiryQ         *Heap
```

Comment:
Why are you adding these two queues (expiryQ and backoffQ) here? This is very confusing.

Reply:
The idea is to treat PodBackoff as a queue for the case of finding the next pod we can schedule so we don't need to perform full pass over all pods we know about in order to determine which are done being 'backed off' (this is backoffQ). Once that work is done its pretty simple to change our GC method to also not require a full pass, so I added expiryQ for this.